Detailed example attribution

ksen0 opened the pull request:
**THIS PR WAS ACCIDENTALLY CLOSED BEFORE COMPLETION. REOPENED HERE: https://github.com/processing/p5.js-website/pull/1009/**

Addresses #415 - work in progress. Remains to be done:

- [x] Support for cases where pre-2023 attribution is also collective [e.g., Shape primitives](https://archive.p5js.org/examples/form-shape-primitives.html)
- [x] Support for cases where there is no link for author
- [ ] Once done on main, will need to be cherrypicked to 2.0 as well
- [ ] Ensure all people mentioned in any part of attribution is on all-contributors in p5.js README in both main and dev-2.0 branches and include "example" emoji. **Complete list:** `Prof** WM Harris`, `GH/prashantgupta24`

# Curent Behavior

The license is at the bottom of the page, and attribution (inspired by and maintained by) is split.

![Screenshot 2025-05-13 at 12 21 45](https://github.com/user-attachments/assets/cfd72917-c53f-47a4-a5d5-0d38dee5db7e)

(Then there is the sketch and the code - viewer has to scroll all the way down)

![Screenshot 2025-05-13 at 12 21 51](https://github.com/user-attachments/assets/c0a84595-ead2-4db9-8e94-ee6ee6583797)

# PR Behavior

Considering showing the attribution more clearly. "p5.js Contributors" also points to all-contributors list in p5.js repo README, rather than the People page. The screenshots includes rephrasing to increase clarity for each example but make the history more maintainable overall using a clearer schema; this is also open to feedback.

![image](https://github.com/user-attachments/assets/2f4883b2-c358-4e39-bccc-f8325da385e2)

![image](https://github.com/user-attachments/assets/82cfaf20-932a-471e-8690-03d40928c94f)

# Future use

Although attribution of p5.js contributors work is collective, this schema allows detailed attribution of referenced work from outside the community. For example, including the following in the `.mdx` of an example:

```
remix:
  - attributionLabel: Processing example by Ira Greenberg
    attributionURL: https://processing.org/examples/reflection1.html
    remixLabel: Based on
  - attributionLabel: David Blitz
    remixLabel: Ported by
  - attributionURL: https://github.com/dkessner
    attributionLabel: Darren Kessner
    remixURL: https://github.com/processing/p5.js-example
    remixLabel: Revised in 2023

collectivelyAttributedSince: 2024
```

Will create the following attribution:

![image](https://github.com/user-attachments/assets/ff547f88-81bc-40af-837a-ff71a5f20714)

Thus, new examples can be added that correctly attribute referenced material. In the above schema, the `collectivelyAttributedSince` field is applicable mainly to examples ported from the Processing or archived p5.js site. For future additions, the `remix` filed can be used to provide a complete history. The URLs for attribution and type of remix can be edited, with optional URLs. If there is no remixLabel, "Remixed from" will be used, but other phrasing can be provided for clarity.

limzykenneth commented on Gruntfile.js:
```diff
@@ -351,6 +352,27 @@ module.exports = function(grunt) {
+    },
+    shell: {
+      make_tmp_dir: {
+        command: 'mkdir -p tmp/p5.js'
```
Is it possible to use `fs` instead of `mkdir` shell command? If possible I would like to avoid using shell commands directly in script as that could cause cross platform compatibility issue.

yukienomiya replied:
The reason why I used shell commands it's that this task is designed to only be run by the GitHub workflow, so I didn't think it was necessary to make it cross-platform compatible.

limzykenneth replied:
In general I would like to reduce reliance on external dependencies if reasonable, but I guess this is not a big issue for me since it will only run in GitHub Actions so it's up to you.

yukienomiya replied:
This might be a dumb question, but why do we want to reduce external dependencies?

limzykenneth replied:
Many external dependencies fall out of maintenance without future support and development from the maintainers, and so the code debt often gets passed down to projects using them. In p5.js the most notables are YUIDoc and arguably Assemble.

Of course it is on a reasonable basis, the fact that YUIDoc is still used is simply because there are no viable drop in replacement, so the aim is not to avoid external dependencies as much as possible but rather to use them cautiously and to use existing dependencies or native node.js functionality when it is not too troublesome to do so or if it affects maintainability.

outofambit replied:
@limzykenneth are you concerned about the external dependency on `mkdir` being available in the environment? i'm happy to file an issue to track using only `fse` as a future improvement, and merge this as-is.

limzykenneth replied:
Yes, the concern is mainly about `mkdir` being available in terms of cross platform-ness. `git` is a bit of a concern as well but not to the same extent. However since this should only run in CI, presumably as long as GitHub Actions support some kind of POSIX environment it should be fine.

limzykenneth replied:
Also for clarification, by external dependencies I mainly mean `grunt-shell` but if we'll go forward using shell script here then that's fine to leave as is.

yukienomiya replied:
@limzykenneth Thanks for the clear explanation 🙏🏻🙏🏻
I can use fs-extra instead of grunt-shell to create the folders (and `simple-git` for the git command)
But I'm still not sure what to do about the `generate_dataJSON` task: should I use grunt-shell to run `npm ci` or something else?

limzykenneth replied:
Unfortunately there isn't an officially supported way to install npm packages programmatically. I've recently looked into this and noticed Parcel does install missing dependencies programmatically. Digging into their code, they are also using command line process spawning (in their case `cross-spawn`).

In our case here we maybe have to go for `grunt-shell` still.

yukienomiya replied:
I see, for now I'll just update the `mkdir` and the git command then.
Thank you for the suggestions! ♥️

limzykenneth commented on Gruntfile.js:
```diff
@@ -351,6 +352,27 @@ module.exports = function(grunt) {
+        command: 'mkdir -p tmp/p5.js'
+      },
+      clone_p5js_repo: {
+        command: 'git clone https://github.com/processing/p5.js .',
```
Same here, I would prefer to avoid calling shell commands directly. You can have a look at the release script in the main p5.js repo and its usage of `simple-git` to run git commands.

limzykenneth commented on Gruntfile.js:
```diff
@@ -351,6 +352,27 @@ module.exports = function(grunt) {
+        }
+      },
+      generate_dataJSON: {
+        command: 'npm i && npm run grunt yui',
```
Perhaps this can run from within grunt itself by calling `grunt.task.run()` with the relevant steps?

yukienomiya replied:
By "this" do you mean the whole grunt-shell command or just the `npm run grunt yui`?

limzykenneth replied:
`npm run grunt yui`, if possible run after install is done. Also for `npm i` it would be better to use `npm ci` since that's what it is made for (running in CI environment)

outofambit commented on .github/workflows/update-documentation.yml:
```diff
@@ -0,0 +1,37 @@
+        with:
+          message: 'Update Reference files for p5.js ${{ steps.p5js.outputs.VERSION }} release (${{ steps.p5js.outputs.SHA }})'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
@yukienomiya i got the secret added!, but i had to use a different key (`GITHUB_TOKEN` must be reserved by GitHub itself or something)
```suggestion
          GITHUB_TOKEN: ${{ secrets.DOCS_COMMIT_TOKEN }}
```

yukienomiya replied:
Thank you! ❤️🎉

ksen0 committed:
Initial format for detailed example attribution

ksen0 committed:
Merge branch 'main' into 415-attrib

calebfoss commented:
@ksen0 This looks great! I really appreciate the thoughtfulness of how you put this together.

One tiny tweak: it occurred to me that since the code was copied over to this repo in 2024, it should be "2024 onwards" rather than 2023.

This certainly does not need to be in this PR, but something to consider for the future: I think it would be helpful for readers to have some visual design distinction between the description and attribution. I'm picturing the ways that text publications design attribution information in articles. Italics is a common convention but may look awkward with this much text. Maybe just a slightly different background color?

Thank you again!

ksen0 committed:
Improved remix attribution schema

ksen0 committed:
Improved remix attribution schema

ksen0 committed:
Merge branch 'main' into 415-attrib

ksen0 committed:
Use specified year for collective attribution where applicable

ksen0 committed:
Use remixLookup to reduce duplicate .mdx headers

ksen0 committed:
Add 2023 revision attributions

ksen0 commented:
Hi @calebfoss ! This is ready for review. I've updated the 2023 to 2024, and created some flexibility for attribution copy in the schema, happy to hear your thoughts.

> Maybe just a slightly different background color?

That seems like a good idea, I will look into that (background color and/or italics), but opening the functionality for review now. Do you think using a slightly smaller font can be an option too or does something speak against it?

Also: I'm not totally sure how to correctly indicate Malay's work in some of the examples (Alpha Mask, Sine Cosine, Aim). Color Wheel and Lyrics Poem does attribute Malay but I wasn't sure on correct `attributionLabel` and `attributionURL` to use.

calebfoss commented:
@ksen0 thank you again for your work on this!

Here are the issues I found:

- Color: missing credit for @dkessner and me
- Conditions: I don't think "remixed" is the best word here, since this shares no code with Prof. Harris's example. Rather, it was intended as a completely new example inspired by the educational intentions of her example.
- Alpha Mask: missing credit for @Malayvasa
- Audio Player: I wrote this one from scratch, but it currently credits @katlich112358 
- Video Capture: missing credit for @katlich112358 
- Translate: missing credit for @dkessner 
- Color Wheel - missing link for @Malayvasa
- Noise - I wrote this one from scratch (missing credit)
- Random Poetry - missing link for @Malayvasa
- Sine and Cosine - missing credit for @Malayvasa 
- Aim - missing credit for @Malayvasa 
- Circle Clicker - I wrote this one from scratch (missing credit)
- Ping Pong - I wrote this one from scratch (missing credit)
- Custom Geometry - @davepagurek wrote this one if he would like credit
- Filter Shader - @davepagurek wrote this one if he would like credit
- Adjusting Positions with a Shader - I think this is @davepagurek too?
- Multiple Canvases - I wrote this one from scratch (missing credit)
- Local Storage - I wrote this one based on the @shiffman JSON and Table examples (missing credit)

Across the examples, none currently link the code on GitHub for the original site version or the 2023 revision version. The 2023 revision link points to the 2023 repo, but not the specific code. As we discussed in #415, linking the specific code from both the p5.js-website-legacy and p5.js-example repos provides a path to the git history, which documents many more contributors' work.

I am really happy to see this getting close to completion! Thanks again!

davepagurek commented:
Those ones you tagged me on are ones I made, yep!

ksen0 commented:
Hi @calebfoss , thanks for the detailed list! While working on this, I'm trying to keep maintenance in mind, too, and the next commit is WIP but below is a question I have in terms of design, if you have any thoughts on the below please let me know!

As an update, I added support for multiple attributions in one remix line ("Revised in 2023 by PersonA, PersonB") so the missing attributions will appear in the next iteration. With that I can cover all the missing credit cases. Not yet committed to PR, but here's an example of `.mdx` syntax in case anyone has input:

```mdx
remix:
  - description: Revised in 2023 by
    attribution:
     - name: Caleb Foss
     - name: Darren Kessner
    codeURL: TODO - specific example code goes here
```

(There's a lookup, `attributionURLLookup` in `config.ts` replacing what's there now but making repeated attributions easier to maintain. If no URL is provided, one from the lookup is used; otherwise, name is not linked. So it's still really flexible.)

I also updated the phrasing or "Remixed from" -> "Inspired by" of the Conditions examples, but for the structure as a whole, from what I understand, "remix" has the widest applicable meaning in modifying and building on open source code in many different ways ([example usage in an analysis of sketches](https://dl.acm.org/doi/pdf/10.1145/3563657.3595969)). Unless there is another similarly flexible alternative, I'd like to keep this usage, at least under the hood. (Because this schema is pretty complex, I'll also be adding some contributor guidelines on using it when this is more defined.)

I've been trying to figure out a way to add the code links to the two repos. Please let me know if the two links in the blow example don't meet the criterion! _(EDITED to reflect slightly renamed fields in an attempt to improve readability)_

```mdx
remix:
  - description: Inspired by
    attribution:
      - name: Prof. WM Harris
        URL: https://people.rit.edu/wmhics/
    codeURL: https://github.com/processing/p5.js-website-legacy/blob/main/src/data/examples/en/04_Control/05_Logical_Operators_2.js
  - description: Revised in 2023 by
    attribution:
      - name: Caleb Foss
    codeURL: https://github.com/processing/p5.js-example/tree/main/examples/02_Animation_And_Variables/03_Conditions
  - collective:
      - since: 2024
```

## Option 1
Less legible because too many links and unclear where the "inspired" and "revised" links go:
![image](https://github.com/user-attachments/assets/b82f4810-7d29-4505-975e-0a240d9fccd6)

## Option 2
More legible, however there are some challenges with this:

![image](https://github.com/user-attachments/assets/0f7294fa-3d32-4c14-addc-4677ebb326a3)

I think if at some point these links become inaccessible, then it's possible to easily identify, remove, or replace with wayback machine versions where possible, so this seems alright, but please let me know if you see any other potential maintainability issues.

Potential issues with option 2:

* Non-descriptive label of "code" link. Possible ideas here would be "code in p5.js-website-legacy repository" but this makes the whole block potentially less readable. (Still better than option 1 above, though, I think?)
* Inconsistency: lack of matching "code" link for the last part that describes current maintenance. Could be added technically, but given the strong concern with maintainability in #415, unless it addresses the specific current attribution problem, then I really hesitate to add it.
 
Anyway, I'm still working on this, and I'll update here if I have a different more compact (in terms of schema and template layout code) solution, but in the meantime I wanted to share in case you had thoughts on this!

calebfoss commented:
Thanks for the update @ksen0!

I'm with you on Option 2 looking like a better direction. The word "code" tells me what to expect by clicking the link. To make the link more clear out of context for accessibility without getting too verbose, perhaps `pre 2023 code`, `2023 revision code`, `post 2023 code`?

Regarding linking the current code, I think of the message included at the bottom of Reference pages:

> Notice any errors or typos? [Please let us know](https://github.com/processing/p5.js/issues). Please feel free to edit `[link to code on GitHub]` and open a pull request!

I would expect this to improve the maintenance process, as it invites the community to help without having to sift through the GitHub repo searching for the matching file. If you had the current GitHub code path in the MDX, you could consider reusing it for both the attribution section and an invitation to help that matches the reference pages.

ksen0 committed:
Updated attribution style

ksen0 committed:
Documented remix attribution schema

ksen0 committed:
Merge branch 'main' into 415-attrib

ksen0 committed:
Remix attribution layout code cleanup

ksen0 committed:
Update examples metadata to use new schema

ksen0 commented:
Here's an update; there's contributor docs in the PR now and as for individual files.

Partial change log:

* Color: added credit for you and dkessner with "Created by" as I didn't see it in the "Curated Examples" spreadsheet, please let me know if there are earlier things to link here?
* Conditions: changed label to "Inspired by"
* Alpha Mask: Added credit in "revised by" segment, because it wasn't clear to me if this one should be "created by" Malay Vasa and Kasey Lichlyter, or created by former and revised by latter?
* Audio Player: updated to "created by Caleb Foss"
* Color Wheel - added link for Malay Vasa
* Random Poetry - added link for Malay Vasa

PR converted to draft, because the following remains: _EDITED Sep 1, 2025 to indicate with ✔️  what's been done_

* Adding all the historical code links (WIP) ✔️ 
* Video Capture: missing credit for Kasey Lichlyter ✔️ 
* Translate: missing credit for @ dkessner ✔️ 
* Noise - missing credit Caleb Foss ✔️ 
* Sine and Cosine - missing credit for @ Malayvasa ✔️ 
* Aim - missing credit for @ Malayvasa ✔️ 
* Circle Clicker -missing credit Caleb Foss ✔️ 
* Ping Pong -missing credit Caleb Foss ✔️ 
* Custom Geometry - @ davepagurek  credit ✔️ 
* Filter Shader - @ davepagurek credit ✔️ 
* Adjusting Positions with a Shader -@ davepagurek  credit ✔️ 
* Multiple Canvases -missing credit Caleb Foss ✔️ 
* Local Storage - missing credit Caleb Foss, based on the @ shiffman JSON and Table examples (missing credit) ✔️ 

> Regarding linking the current code, I think of the message included at the bottom of Reference pages:

In principle this is a good idea, though the scope of this PR is already a bit large. In general, on some more reflection, I think it would be good to work toward a general "edit this page" functionality and that's something that's already on the radar, though may be technically pretty challenging. The reference specifically is different enough in how it's generated that I don't thik that can be used directly. So I think this should be left out of the scope of work here. As shown in examples below, I've included a placeholder that can be improved in the future.

> pre 2023 code, 2023 revision code, post 2023 code?

This is roughly the approach, though it is manual. I tried a few different configurations and feels like the below is going in the right direction. Some examples - showing that the past code is made available with legible labels (with a placeholder for the "current code" that can be improved later), right after a readable and reusable attribution segment:

<img width="805" height="157" alt="Screenshot 2025-07-28 at 13 43 56" src="https://github.com/user-attachments/assets/054205c5-6306-4672-86e2-e24e9a43d515" />

<img width="832" height="210" alt="Screenshot 2025-07-28 at 13 13 59" src="https://github.com/user-attachments/assets/feeb0aeb-7f89-44d0-ac63-faab144cc3c5" />

<img width="795" height="481" alt="Screenshot 2025-07-28 at 13 35 35" src="https://github.com/user-attachments/assets/b9bb1af3-3278-4ff9-b5d0-27a90165eb9b" />

The code list here is separated and flattened for legibility - not sure if this really works, but the many parentheses were very difficult to navigate. However, the underlying schema still has these lists associated with specific "remix" entries, so other layouts can definitely be tried here too - this is just the best I could come up with based on both the design/copy ideas above, and the constraints of making this schema flexible for all the current cases and reusable in the future.

_EDIT: Strings should also be localized, for example below is what is shown when switching to another language:_

<img width="832" height="498" alt="image" src="https://github.com/user-attachments/assets/c043a987-f584-4147-b56d-3dac2d2ca403" />

calebfoss commented:
@ksen0 Anything I can do to help push this across the finish line?

ksen0 committed:
Examples attribution metadata

ksen0 committed:
Merge branch 'main' into 415-attrib

ksen0 committed:
Examples attribute metadata - additional fixes

ksen0 commented:
Hi @calebfoss , thanks for bumping this up, here's an update. The last commits should already have all the example info; I tried to go through all the prior lists, but please do let me know what needs to be added/changed. (I also didn't see any specific attribution detail to be added to https://p5js.org/examples/calculating-values-map/ or https://p5js.org/examples/calculating-values-interpolate/ - is that correct, or did I miss something?)

However, what remains is also making this localizable. Even if there are not hi/ko/es/zh-Hans strings available, the layout code right now hard-codes English phrases, as does the schema - so this has to be changed to use translated strings, or en as a fallback, the same way the rest of the site does. So now I'm working on that. If in the meantime you have any feedback about the attributions details, please let me know? (Links, names, name order, labels - like "Created by / Inspired by" etc, which would be localizable in the next iteration - the contributor docs for this, etc.)

_EDIT Sep 29: Sorry for the delays again; I think I should be able to wrap this up this week. Last week I also got some feedback on the usability/readability of these texts, so I'll add some minor tweaks, but main task is still the internationalization (not translation itself, but correctly using the localizable strings)_

calebfoss commented:
@ksen0 Thanks for the updates, and I am so glad to hear you are wrapping this up this week! I can't tell you how much of a relief that will be.

I haven't had a chance to run the latest version locally, but I wanted to reply about the Map and Interpolate examples. Both of those are revisions of examples from the old site ([Map](https://github.com/processing/p5.js-website-legacy/blob/main/src/data/examples/en/08_Math/19_Map.js) and [Lerp](https://github.com/processing/p5.js-website-legacy/blob/main/src/data/examples/en/08_Math/10_Interpolate.js)) and I was the editor on the 2023 revision.

Regarding language around "created by", "inspired by", etc., I'll try to find time to look through the latest version and share feedback, but the main priority, of course, is getting the creators' names added/linked. 

When we were working on this project, it became clear that categorizing different forms of contribution on these would be awkward/complicated, given that small changes can have a big impact, and in some cases we were completely rewriting examples but still wanted to pay respect to the creative work that led us to the new versions. That is why we recommended simply listing everyone who had contributed to each example, recognizing the collaboration that occurred over the course of years to make each example what it is. 

Language like "created by" or "inspired by" is going to be an inherently imperfect way to describe that process, but I have been doing my best to help with this approach and appreciate the attention to detail you have invested.

calebfoss commented on src/content/examples/en/01_Shapes_And_Color/00_Shape_Primitives/description.mdx:
```diff
@@ -6,6 +6,24 @@ oneLineDescription: Draw 2D shapes.
+    attribution:
+     - name: Darren Kessner
+       URL: https://github.com/dkessner
+  - description: Created by
```
This current credits Darren and me for creating it, and credits Darren for revising it. This reads a bit confusing, especially since this was a rewrite of [the example from the old site](https://archive.p5js.org/examples/form-shape-primitives.html) (my memory is that it shares no specific code but has the same goal. Since the old site did not credit a creator, and Darren did more work on this than me, maybe just credit Darren as "Rewritten by"?

calebfoss commented on src/content/examples/en/02_Animation_And_Variables/03_Conditions/description.mdx:
```diff
@@ -3,6 +3,24 @@ featuredImage: "../../../images/featured/02_Animation_And_Variables-03_Condition
 oneLineDescription: Use if and else statements to make decisions while your sketch runs.
+
+remix:
+  - description: Inspired by
```
The grammar here seems to communicate that the example was inspired by Prof. Harris herself. I think most readers would figure out that it's really referring to something she created, but ideally this would read something like:

Inspired by [the Conditional Shapes example](https://archive.p5js.org/examples/control-conditional-shapes.html) by [Prof. WM Harris](https://people.rit.edu/wmhics/)

calebfoss commented on src/content/examples/en/04_Input_Elements/02_DOM_Form_Elements/description.mdx:
```diff
@@ -3,6 +3,22 @@ featuredImage: "../../../images/featured/04_Input_Elements-02_DOM_Form_Elements-
+remix:
+  - attribution:
+      - URL: https://people.rit.edu/wmhics/
+        name: Prof. WM Harris
```
Right now this ends up rendering the text as

> Remixed by [Prof. WM Harris](https://people.rit.edu/wmhics/)

Looks like it's just missing the description field, where in this case I think "Created by" would be appropriate, since the new example does contain a significant amount of code from the original.

calebfoss commented on src/content/examples/en/07_Repetition/05_Recursive_Tree/description.mdx:
```diff
@@ -3,6 +3,30 @@ featuredImage: "../../../images/featured/07_Repetition-05_Recursive_Tree-thumbna
+
+
+remix:
+  - description: Inspired by
```
This is case like another I point out in a comment in which the literal grammar reads that this was inspired by Dan Shiffman himself, rather than his example. Again, I think people can figure it out, and in this case it's clarified in the description. I wanted to note it, though, in case you want to change it on all relevant examples.

calebfoss commented on src/content/examples/en/11_3D/05_Adjusting_Positions_With_A_Shader/description.mdx:
```diff
@@ -3,6 +3,15 @@ featuredImage: "../../../images/featured/11_3D-05_Adjusting_Positions_With_A_Sha
 oneLineDescription: Use a shader to adjust shape vertices.
+
+remix:
+  - description: Creared by
```
Typo `Crearer` instead of `Created`

calebfoss commented on src/content/examples/en/14_Loading_And_Saving_Data/00_Local_Storage/description.mdx:
```diff
@@ -3,6 +3,26 @@ featuredImage: "../../../images/featured/14_Loading_And_Saving_Data-00_Local_Sto
+    code:
+      - label: 2023 code
+        URL: https://github.com/processing/p5.js-example/tree/main/examples/14_Loading_And_Saving_Data/00_Local_Storage
+  - description: Inspired by
```
Another inspired by a human example (see other comments)

calebfoss commented:
@ksen0 - Okay, I finally ran this, and this time I left comments directly on the MDX files, which I hope makes things easier. Thank you again for all your work on this. I am really excited to see the documentation of years of collaboration back on the site, with more consistency than the old site.

ksen0 commented:
Reopened here: https://github.com/processing/p5.js-website/pull/1009/ all comments from above will be actioned
